nixos/cgit: init

schnusch opened the pull request:
###### Description of changes

Add cgit behind nginx using fcgiwrap.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

schnusch committed:
nixos/cgit: init

Lassulus commented:
How is this different from https://github.com/NixOS/nixpkgs/pull/110270 ?

Ma27 commented:
Also, what's wrong with `services.cgit`?

schnusch commented:
I put it at `services.nginx.cgit` to make it consistent with `services.nginx.gitweb` or `services.lighttpd.cgit`.

Ma27 commented:
OTOH we have a lot of other services that are not part of `services.{nginx,httpd}`, but configure vhosts for these reverse proxies. IMHO the `services.nginx` namespace should only be used for options that are **specific** to the nginx service and not for other services that happen to use nginx.

schnusch commented:
I added support for a location prefix. This should now be a -- I think -- cleaner version of #110270, with the order of options in the configuration file, fastcgiParams, and git clone support.

I moved the options to `services.nginx.virtualHost.….cgit`, this allows one to run cgit on multiple vhosts.

Ma27 commented:
I'd like to repeat that it's IMO a bad idea to put options for a service under the nginx namespace.

schnusch commented:
I understand, but if multiple parallel instances of cgit are desired, I think, it's the best place to enable them.

Ma27 commented:
why is it necessary to (ab)use the nginx namespace for that purpose though? this can also be implemented under e.g. the namespace services.cgit.

schnusch commented:
I moved it all to `services.cgit.*`. `services.cgit` is a list now, which seems uncommon in nixpkgs, but allows us to configure multiple instances of cgit.

Lassulus commented on nixos/modules/services/networking/cgit.nix:
```diff
@@ -0,0 +1,192 @@
+{
+  options = {
+    services.cgit = mkOption {
+      type = types.listOf (types.submodule ({ config, ... }: {
```
why not use types.attrsOf? with they key being the default value for nginx.virtualHost?

schnusch replied:
This way currently gives the ability to run multiple cgit instances on the same vhost.

Lassulus replied:
you could do the same with attrsOf.
```
services.cgit."git.example.com" = { ... };
services.cgit."git.example.com2" = { nginx.virtualHost = "git.example.com"; nginx.location = "/git"; }
```

also I'm not sure if having multiple cgits on one instance is a common usecase.
attrs are generally preferred because they can be referenced again at some other point or can be overridden which is not possible with lists

Lassulus commented on nixos/modules/services/networking/cgit.nix:
```diff
@@ -0,0 +1,194 @@
+          nginx.virtualHost = mkOption {
+            description = "VirtualHost to serve cgit on, defaults to the attribute name.";
+            type = types.nullOr types.str;
+            default = null;
```
you can use `default = default = config._module.args.name;` to make the default the attribute key of the attrsOf

Lassulus commented on nixos/tests/cgit.nix:
```diff
@@ -0,0 +1,58 @@
+    server.fail("curl -fsS http://localhost/robots.txt")
+
+    server.succeed(
+        "git init --bare -b master /srv/git/some-repo"
+        " && git init -b master reference"
+        " && (cd reference"
+        " && git remote add origin /srv/git/some-repo"
+        " && date > date.txt"
+        " && git add date.txt"
+        " && git -c user.name=test -c user.email=test@localhost commit -m 'add date'"
+        " && git push -u origin master)"
```
maybe something like this would be cooler?
```
"${pkgs.writeScript "test-git" ''
  #!/bin/sh
  set -efu
  git init --bare -b matser /srv/git/some-repo
  git init -b master reference
  ...
''}"
```

Lassulus reviewed (approved):
Alright, guess I'm happy with it, do you want to squash the changes? or should I do it when merging?

yrd commented on nixos/modules/services/networking/cgit.nix:
```diff
@@ -0,0 +1,201 @@
+            [ "cgit.css" "cgit.png" "favicon.ico" "robots.txt" ]
+            (name: nameValuePair "= ${stripLocation cfg}/${name}" {
+              extraConfig = ''
+                alias ${pkgs.cgit}/cgit/${name};
```
Thanks for your work! Could you please add a `package` option to the module so it's easier to use a custom cgit build?

Ma27 replied:
What's wrong with overlays?

yrd replied:
Nothing - just personal preference, really. It would however allow someone to use two different builds under different virtual hosts, although I have to admit I can't really come up with a reason why someone would need that…

schnusch committed:
nixos/cgit: add package option

yrd reviewed (approved):
@schnusch Thank you!

SuperSandro2000 commented on nixos/modules/services/networking/cgit.nix:
```diff
@@ -0,0 +1,204 @@
+            description = mdDoc "A path which will be scanned for repositories.";
+            type = types.nullOr types.path;
+            default = null;
+            example = "/srv/git";
```
```suggestion
            example = "/var/lib/git";
```
Nixos sues /var/lib for data

SuperSandro2000 commented on nixos/modules/services/networking/cgit.nix:
```diff
@@ -0,0 +1,204 @@
+            type = with types; attrsOf (attrsOf settingType);
+            default = {};
+            example = {
+              blah = {
+                path = "/srv/git/blah";
+                desc = "some repo";
+              };
```
```suggestion
              example = {
                path = "/var/lib/git/example";
                desc = "An example repository";
              };
```

SuperSandro2000 reviewed (commented):
Please rebase that we in the end only have the init commit left or should we squash merge?

yrd commented on nixos/modules/services/networking/cgit.nix:
```diff
@@ -0,0 +1,203 @@
+        options = {
+          enable = mkEnableOption (mdDoc "cgit");
+
+          package = mkPackageOptionMD pkgs "cgit" {};
```
This line fails for me (`error: undefined variable 'mkPackageOptionMD'`). Am I missing something here?

schnusch replied:
I added it in <https://github.com/NixOS/nixpkgs/pull/167319#event-8398250015>, because the doc wouldn't build. 9da5f12ecffa60f7a4712516177ede318b608f0d of #208407 added that.

yrd replied:
I see, thanks!

softinio commented:
As this is approved, any chance it can be merged? Thank you!
